src/openrct2/core/Crypt.OpenRCT2.cpp

```diff
@@ -7,190 +7,95 @@
-        }
-
-        /* Main loop */
-        a = H[0];
-        b = H[1];
-        c = H[2];
-        d = H[3];
-        e = H[4];
-
-        for (idx = 0; idx <= 79; idx++)
-        {
-            if (idx <= 19)
-            {
-                f = (b & c) | ((~b) & d);
-                k = 0x5A827999;
-            }
-            else if (idx >= 20 && idx <= 39)
-            {
-                f = b ^ c ^ d;
-                k = 0x6ED9EBA1;
-            }
-            else if (idx >= 40 && idx <= 59)
-            {
-                f = (b & c) | (b & d) | (c & d);
-                k = 0x8F1BBCDC;
-            }
-            else if (idx >= 60 && idx <= 79)
-            {
-                f = b ^ c ^ d;
-                k = 0xCA62C1D6;
-            }
-            temp = SHA1ROTATELEFT(a, 5) + f + e + k + W[idx];
-            e = d;
-            d = c;
-            c = SHA1ROTATELEFT(b, 30);
-            b = a;
-            a = temp;
-        }
+private:
+    static constexpr uint64_t Offset = 0xCBF29CE484222325ULL;
+    static constexpr uint64_t Prime = 0x00000100000001B3ULL;
 
-        H[0] += a;
-        H[1] += b;
-        H[2] += c;
-        H[3] += d;
-        H[4] += e;
-    }
+    uint64_t _data = Offset;
+    uint8_t _rem[8]{};
+    size_t _remLen{};
 
-    /* Store binary digest in supplied buffer */
-    if (digest)
+    void ProcessRemainder()
     {
-        for (idx = 0; idx < 5; idx++)
+        if (_remLen > 0)
         {
-            digest[idx * 4 + 0] = static_cast<uint8_t>(H[idx] >> 24);
-            digest[idx * 4 + 1] = static_cast<uint8_t>(H[idx] >> 16);
-            digest[idx * 4 + 2] = static_cast<uint8_t>(H[idx] >> 8);
-            digest[idx * 4 + 3] = static_cast<uint8_t>(H[idx]);
+            uint64_t temp{};
+            std::memcpy(&temp, _rem, _remLen);
+            _data ^= temp;
+            _data *= Prime;
+            _remLen = 0;
         }
     }
-    return 0;
-}
-
-class OpenRCT2Sha1Algorithm : public Sha1Algorithm
-{
-private:
-    std::vector<uint8_t> _data;
 
 public:
     HashAlgorithm* Clear() override
     {
-        _data = {};
+        _data = Offset;
         return this;
     }
 
     HashAlgorithm* Update(const void* data, size_t dataLen) override
     {
-        auto src = reinterpret_cast<const uint8_t*>(data);
-        _data.insert(_data.end(), src, src + dataLen);
+        if (dataLen == 0)
+            return this;
+
+        auto src = reinterpret_cast<const uint64_t*>(data);
+        if (_remLen > 0)
+        {
+            // We have remainder, so fill rest of it with bytes from src
+            auto fillLen = sizeof(uint64_t) - _remLen;
+            assert(_remLen + fillLen <= sizeof(uint64_t));
+            std::memcpy(_rem + _remLen, src, fillLen);
+            src = reinterpret_cast<const uint64_t*>(reinterpret_cast<const uint8_t*>(src) + fillLen);
+            _remLen += fillLen;
+            dataLen -= fillLen;
+            ProcessRemainder();
+        }
```

Comment:
Why is this needed? The loop of mine copies just the bytes into uint64_t, if its less than 8 bytes its null padded.

Reply:
Would your code cope with:
  Update(5 bytes)
  Update(3 more bytes)

and give same result as if you did Update(all 8 bytes)?

Reply:
No. A lot of hash algorithms that use blocks will not do this.
https://godbolt.org/z/1875K31nG

There is also really no reason to have this, this is not a cryptographically secure hash, they also can have different schemas.

Reply:
> There is also really no reason to have this

Sorry I disagree with this. The API is so that you can stream the data into the hash algorithm. It should not give you different results depending on the size of the chunks you feed it.

Reply:
How would you implement lets say Murmur2 like that?

Reply:
Exactly the same way.

Reply:
I mean okay, it doesn't matter too much. The thing that sticks out the most is reinterpret_cast, you should avoid it when possible, since we are not on C++20 yet there is no std::bit_cast so the std::memcpy will have to do.